Fix Compiling from `nargo` on M2 Mac

Savio-Sou opened the pull request:
This fixes the errors encountered when `cargo install`ing nargo on M2 (which goes through build.rs).

Assume `libomp` installed with `brew`.

Savio-Sou committed:
Fix Compiling on M1/M2 Macs

Assume `libomp` installed with `brew`.

jfecher commented on barretenberg_wrapper/build.rs:
```diff
@@ -219,12 +220,11 @@ fn link_lib_omp() {
     if let OS::Linux = select_os() {
         let llvm_dir = find_llvm_linux_path();
         println!("cargo:rustc-link-search={}/lib", llvm_dir);
+    } else if let ARM_APPLE = select_toolchain() {
```
Since we're just checking string equality we don't need to use pattern matching here. Should be `else if ARM_APPLE == select_toolchain()` or the reverse `select_toolchain() == ARM_APPLE`

Edit: Looks like you wrote it this way because of the OS::Linux / ARM_LINUX lines above/below, I'd prefer if both were updated

Savio-Sou replied:
Good point. Rewrote matching in the new commit though to include Homebrew path for Intel Macs.

Savio-Sou committed:
Update Toolchain Checks

- Add Intel Mac Homebrew path
- Optimise toolchain equality checks / matching
- Reformat comments

Savio-Sou commented on barretenberg_wrapper/build.rs:
```diff
@@ -216,17 +210,18 @@ fn main() {
+            let llvm_dir = find_llvm_linux_path();
+            println!("cargo:rustc-link-search={}/lib", llvm_dir)
+        }
+        INTEL_APPLE => println!("cargo:rustc-link-search=/usr/local/lib"),
```
Add Intel Mac Homebrew path

Savio-Sou commented on barretenberg_wrapper/build.rs:
```diff
@@ -105,8 +100,8 @@ fn main() {
-    // Link lib OMP
-    link_lib_omp();
+    // Link lib OpenMP
+    link_lib_omp(toolchain);
```
Pass retrieved `toolchain` down to minimise runs of `select_toolchain()`.

Savio-Sou commented on barretenberg_wrapper/build.rs:
```diff
@@ -33,14 +31,14 @@ fn select_arch() -> Arch {
     let arch = std::env::consts::ARCH;
     match arch {
         "arm" => Arch::Arm,
+        "aarch64" => Arch::Arm,
```
This helps M2 Macs with returned architecture of `aarch64` (instead of `arm`) to pick the correct toolchain.

It prevents them running on the fallback x86_64 config that brings compilation involving assembly code, which would lead to the `unknown token in expression` error when ran on ARM (described in more details [here](https://github.com/AztecProtocol/barretenberg/issues/66)).

Savio-Sou commented on barretenberg_wrapper/build.rs:
```diff
@@ -215,18 +210,18 @@ fn main() {
+            println!("cargo:rustc-link-search={}/lib", llvm_dir)
+        }
+        INTEL_APPLE => println!("cargo:rustc-link-search=/usr/local/lib"),
+        ARM_APPLE => println!("cargo:rustc-link-search=/opt/homebrew/lib"),
```
This fixes the error from `clang` on macOS failing to find `lomp` when no search path is provided:
```
error: linking with `cc` failed: exit status: 1
  = note: ld: library not found for -lomp
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
